include/zenoh-pico/api/primitives.h

```diff
@@ -436,6 +423,36 @@ const z_loaned_bytes_t *z_reply_err_payload(const z_loaned_reply_err_t *reply_er
  */
 const z_loaned_encoding_t *z_reply_err_encoding(const z_loaned_reply_err_t *reply_err);
 
+/**
+ * Builds a :c:type:`z_owned_slice_t` by copying a buffer into it.
+ *
+ * Parameters:
+ *   slice: Pointer to an uninitialized :c:type:`z_owned_slice_t`.
+ *   data: Pointer to the data that will be copied into slice.
+ *   len: Number of bytes to copy.
+ *
+ * Return:
+ *   ``0`` if creation successful, ``negative value`` otherwise.
+ */
+int8_t z_slice_from_buf(z_owned_slice_t *slice, const uint8_t *data, size_t len);
+
+/**
+ * Builds a :c:type:`z_owned_slice_t` by transferring ownership over a data to it.
+ *
+ * Parameters:
+ *   slice: Pointer to an uninitialized :c:type:`z_owned_slice_t`.
+ *   data: Pointer to the data to be owned by `slice`.
+ *   len: Number of bytes in `data`.
+ *  deleter: A thread-safe delete function to free the `data`. Will be called once when `slice` is dropped. Can be
+ * NULL, in case if `data` is allocated in static memory.
+ *    context: An optional context to be passed to the `deleter`.
+ *
+ * Return:
+ *   ``0`` if creation successful, ``negative value`` otherwise.
+ */
+int8_t z_slice_wrap(z_owned_slice_t *slice, uint8_t *data, size_t len, void (*deleter)(void *data, void *context),
```

Comment:
The parameter have type `uint8_t*' but deleter accepts `void*`. This makes not obvious that deleter accepts the same data which is passed to the function. Maybe deleters should accept the same type which is passed to function.

Counterargument may be that this makes harder to make common deleter for all functions. I agree with that. One solution, which may also improve the usability could be like this:
```C
z_payload_from_str(z_owned_payload_t* payload, char* str, void (*free)(void* str));
z_payload_from_str_with_ctx(z_owned_payload_t* payload, char* str, void (*free)(char* str, void* ctx), void* ctx);
z_payload_from_static_str(z_owned_payload_t* payload, const char* str);
```
This will allow user in most cases simply call
```C
z_owned_payload_t payload;
z_payload_from_str(&payload, strdup("foo"), free);
```
without creating any additonal release function.

And if user have some more complex allocators, he have "_ctx" functions for it accepting deleters, which should be implemented manually anyway.

Reply:
If we accept typed deleter we will have to always wrap it in void* deleter (since strings are implemented in terms of slice in both zenoh-c and pico), which is an unnecessary extra cost.
If the user use strdup, they could simply call copy-making serialization function. Otherwise there is no common free function in C (even the one used in zenoh-c is likely different from standard free since it uses rust Box allocation/ deallocation), so I do not think it is makes much sense to bloat the api with functions that would likely lead to errors and a lot confusion and frustration (it is pretty difficult to debug the fact that wrong free is being called).